manifests/resources/web.pp

```diff
@@ -13,38 +13,32 @@
 # for loading this class manually.
 #
 class zabbix::resources::web (
-  $zabbix_url     = undef,
-  $zabbix_user    = undef,
-  $zabbix_pass    = undef,
-  $apache_use_ssl = undef,
+  String[1] $zabbix_url,
+  String[1] $zabbix_user,
+  String[1] $zabbix_pass,
+  Boolean   $apache_use_ssl,
 ) {
+  file { '/etc/zabbix/api.conf':
+    ensure  => file,
+    owner   => 'root',
+    group   => 'root',
+    mode    => '0400',
+    content => epp('zabbix/api.conf.epp', {
+      zabbix_url     => $zabbix_url,
+      zabbix_user    => $zabbix_user,
+      zabbix_pass    => $zabbix_pass,
+      apache_use_ssl => $apache_use_ssl,
+    }),
+  }
 
   Zabbix_proxy <<| |>> {
-    zabbix_url     => $zabbix_url,
-    zabbix_user    => $zabbix_user,
-    zabbix_pass    => $zabbix_pass,
-    apache_use_ssl => $apache_use_ssl,
     require        => [
       Service['zabbix-server'],
       Package['zabbixapi'],
+      File['/etc/zabbix/api.conf'],
     ],
   }
-  -> Zabbix_template <<| |>> {
-    zabbix_url     => $zabbix_url,
-    zabbix_user    => $zabbix_user,
-    zabbix_pass    => $zabbix_pass,
-    apache_use_ssl => $apache_use_ssl,
-  }
-  -> Zabbix_host <<| |>> {
-    zabbix_url     => $zabbix_url,
-    zabbix_user    => $zabbix_user,
-    zabbix_pass    => $zabbix_pass,
-    apache_use_ssl => $apache_use_ssl,
-  }
-  -> Zabbix_userparameters <<| |>> {
-    zabbix_url     => $zabbix_url,
-    zabbix_user    => $zabbix_user,
-    zabbix_pass    => $zabbix_pass,
-    apache_use_ssl => $apache_use_ssl,
-  }
+  -> Zabbix_template <<| |>>
```

Comment:
Are these needed anymore?  Perhaps they are redundant (given the types all autorequire the API config file)?

It depends on whether they were just being used to set the API credentials or whether actually eg. all Zabbix_userparameters have to come after all Zabbix_hosts etc?

> When given a block of attributes and values, a collector will set and override those attributes for every resource (virtual or not) that matches its search expression.
https://puppet.com/docs/puppet/5.3/lang_resources_advanced.html#amending-attributes-with-a-collector

Reply:
From my understanding they ensure that e.g. the host exists in Puppet before the template gets added to it.

Reply:
Fair enough.  Just checking that the relationships before this change were actually all required (and not just a slightly lazy syntax short-cut).  Now they definitely look like they're there for a reason, so it's good to check.